### PR TITLE
Credentials class refactor

### PR DESCRIPTION
**Pull request by oleurud**
Refactor of the Credentials class

The first step of https://github.com/CartoDB/cartoframes/issues/661

In order to avoid a huge PR review like the last one, I am going to create several like this one. The objective is to allow you to check the structure or the idea since the tests are not going to be green. Then, in the main PR (using `feature/credentials-refactor`) we will see all together.

If you see a better way, please let me know

### CODE REVIEW

File: cartoframes/auth/credentials.py [hunk 1]
```diff
@@ -23,79 +23,111 @@ class Credentials(object):
...
+        """Retrives credentials from another Credentials object"""
+        if isinstance(credentials, Credentials):
+            return cls(credentials.api_key, credentials.username, credentials.base_url, credentials.session)
+
+        raise ValueError('`credentials` must a Credentials class instance')
 
     def __repr__(self):
         return ('Credentials(username={username}, '
-                'key={key}, '
+                'api_key={api_key}, '
```
Comment (andy-esch) [inline cartoframes/auth/credentials.py]: It's my fault, but could we put the items in quotes so the code is reproducible? 

```python
Credentials(username='{username}', api_key='{api_key}')
```

Comment (oleurud) [reply cartoframes/auth/credentials.py]: Of course :)

File: cartoframes/auth/credentials.py [hunk 2]
```diff
@@ -23,79 +23,111 @@ class Credentials(object):
...
+
+        self._api_key = api_key
+        self._username = username
+        self.base_url = base_url or 'https://{}.carto.com/'.format(self._username)
+        self._session = session
+
+        self._norm_credentials()
+
+    @classmethod
+    def create_from_file(cls, config_file=None):
```
Comment (andy-esch) [inline cartoframes/auth/credentials.py]: I like the `create_` part of the classmethod name, but also feel like it's implicitly understood that this method creates a credentials file. Overall, I think I support removing it and naming it `from_file(...)` instead. I believe that's the 'standard'. It also requires the user to write seven fewer characters.

Comment (andy-esch) [reply cartoframes/auth/credentials.py]: Same for the other class methods with similar names

Comment (oleurud) [reply cartoframes/auth/credentials.py]: Yeah, I agree. I always use many characters

**Comment by oleurud** [comment]
@andy-esch Your work was and is awesome! This is the easy part

File: cartoframes/auth/credentials.py [hunk 3]
```diff
@@ -23,79 +23,111 @@ class Credentials(object):
...
+
+        return cls(credentials.get('api_key'), credentials.get('username'))
+
+    @classmethod
+    def from_credentials(cls, credentials):
+        """Retrives credentials from another Credentials object"""
+        if isinstance(credentials, Credentials):
+            return cls(credentials.api_key, credentials.username, credentials.base_url, credentials.session)
+
+        raise ValueError('`credentials` must a Credentials class instance')
```
Comment (alrocar) [inline cartoframes/auth/credentials.py]: typo -> `must be` 👼

File: cartoframes/auth/credentials.py [hunk 4]
```diff
@@ -23,79 +23,111 @@ class Credentials(object):
...
+
+        self._norm_credentials()
+
+    @classmethod
+    def from_file(cls, config_file=None):
+        """Retrives credentials from a file. Defaults to the user config directory"""
+        with open(config_file or _DEFAULT_PATH, 'r') as f:
+            credentials = json.load(f)
+
+        return cls(credentials.get('api_key'), credentials.get('username'))
```
Comment (alrocar) [inline cartoframes/auth/credentials.py]: Also this makes the saved credentials not to work in on-premises environments, right?

Comment (oleurud) [reply cartoframes/auth/credentials.py]: Good catch, it is a bug

**Comment by oleurud** [comment]
It is not late! We can do the stuff in the "main PR"

File: cartoframes/auth/credentials.py (hunk 3 above)
Comment (oleurud) [reply cartoframes/auth/credentials.py]: Done in https://github.com/CartoDB/cartoframes/pull/850

File: cartoframes/auth/credentials.py (hunk 4 above)
Comment (oleurud) [reply cartoframes/auth/credentials.py]: Done in https://github.com/CartoDB/cartoframes/pull/850

Comment (andy-esch) [reply cartoframes/auth/credentials.py]: Great catch!!

**Pull request merged**